File: savo_ws/src/core/savo_head/savo_head/drivers/pantilt_driver.py
# -*- coding: utf-8 -*-

"""Python fallback pan-tilt driver for Robot Savo head diagnostics."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

from savo_head.constants import (
    HEAD_BACKEND_DEFAULT,
    I2C_BUS_DEFAULT,
    PCA9685_ADDR_DEFAULT,
    PCA9685_PWM_FREQ_HZ_DEFAULT,
    STATUS_DRYRUN,
    STATUS_OK,
)
from savo_head.core.calibration import (
    DEFAULT_SERVO_CALIBRATION,
    HeadServoCalibration,
    ServoChannelCalibration,
    angle_to_pulse_us,
    calibration_for_axis,
)
from savo_head.drivers.pca9685_driver import Pca9685Config, Pca9685Driver
from savo_head.models.head_state import (
    MODE_CENTERING,
    MODE_IDLE,
    MODE_MANUAL,
    PanTiltLimits,
    PanTiltState,
)
from savo_head.models.pantilt_command import (
    COMMAND_STOP,
    PanTiltCommand,
    center_command,
    hold_command,
)


BACKEND_PCA9685 = "pca9685"
BACKEND_DRYRUN = "dryrun"


@dataclass(frozen=True)
class ServoOutput:
    axis: str
    logical_channel: str
    pca9685_channel: int
    angle_deg: int
    pulse_us: float
    ticks: int

    def to_dict(self) -> dict:
        return {
            "axis": self.axis,
            "logical_channel": self.logical_channel,
            "pca9685_channel": self.pca9685_channel,
            "angle_deg": self.angle_deg,
            "pulse_us": self.pulse_us,
            "ticks": self.ticks,
        }


@dataclass(frozen=True)
class PanTiltDriverConfig:
    backend: str = HEAD_BACKEND_DEFAULT
    i2c_bus: int = I2C_BUS_DEFAULT
    pca9685_address: int = PCA9685_ADDR_DEFAULT
    pwm_frequency_hz: float = PCA9685_PWM_FREQ_HZ_DEFAULT
    calibration: HeadServoCalibration = field(default_factory=lambda: DEFAULT_SERVO_CALIBRATION)
    center_on_open: bool = False
    center_on_close: bool = False

    def normalized(self) -> "PanTiltDriverConfig":
        backend = str(self.backend).strip().lower()
        if backend not in (BACKEND_PCA9685, BACKEND_DRYRUN):
            backend = BACKEND_DRYRUN

        return PanTiltDriverConfig(
            backend=backend,
            i2c_bus=int(self.i2c_bus),
            pca9685_address=int(self.pca9685_address),
            pwm_frequency_hz=float(self.pwm_frequency_hz),
            calibration=self.calibration.normalized(),
            center_on_open=bool(self.center_on_open),
            center_on_close=bool(self.center_on_close),
        )

    def dryrun(self) -> bool:
        return self.normalized().backend == BACKEND_DRYRUN


class PanTiltDriver:
    def __init__(
        self,
        config: PanTiltDriverConfig | None = None,
        pca_driver: Optional[Pca9685Driver] = None,
    ):
        self.config = (config or PanTiltDriverConfig()).normalized()
        self.calibration = self.config.calibration.normalized()

        self._pca_driver = pca_driver or Pca9685Driver(
            Pca9685Config(
                i2c_bus=self.config.i2c_bus,
                address=self.config.pca9685_address,
                pwm_frequency_hz=self.config.pwm_frequency_hz,
                dryrun=self.config.dryrun(),
            )
        )

        self._opened = False
        self._state = PanTiltState(
            pan_deg=self.calibration.pan.center_deg,
            tilt_deg=self.calibration.tilt.center_deg,
            mode=MODE_IDLE,
            status=STATUS_DRYRUN if self.config.dryrun() else STATUS_OK,
            source="init",
        )
        self._last_outputs: list[ServoOutput] = []

    @property
    def opened(self) -> bool:
        return self._opened

    @property
    def state(self) -> PanTiltState:
        return self._state

    @property
    def last_outputs(self) -> tuple[ServoOutput, ...]:
        return tuple(self._last_outputs)

    def open(self) -> None:
        if self._opened:
            return

        self._pca_driver.open()
        self._opened = True

        if self.config.center_on_open:
            self.center(source="open")

    def close(self) -> None:
        if self._opened and self.config.center_on_close:
            self.center(source="close")

        self._pca_driver.close()
        self._opened = False

    def ensure_open(self) -> None:
        if not self._opened:
            self.open()

    def limits(self) -> PanTiltLimits:
        return PanTiltLimits(
            pan_min_deg=self.calibration.pan.min_deg,
            pan_center_deg=self.calibration.pan.center_deg,
            pan_max_deg=self.calibration.pan.max_deg,
            tilt_min_deg=self.calibration.tilt.min_deg,
            tilt_center_deg=self.calibration.tilt.center_deg,
            tilt_max_deg=self.calibration.tilt.max_deg,
        )

    def _servo_output(self, axis: str, angle_deg: int) -> ServoOutput:
        item = calibration_for_axis(axis, self.calibration)
        angle = item.clamp_angle(angle_deg)
        pulse_us = angle_to_pulse_us(
            angle,
            error_deg=item.error_deg,
            direction=item.direction,
        )
        ticks = self._pca_driver.set_servo_pulse_us(item.pca9685_channel, pulse_us)

        return ServoOutput(
            axis=axis,
            logical_channel=item.logical_channel,
            pca9685_channel=item.pca9685_channel,
            angle_deg=angle,
            pulse_us=pulse_us,
            ticks=ticks,
        )

    def set_axis(
        self,
        axis: str,
        angle_deg: int,
        *,
        mode: str = MODE_MANUAL,
        source: str = "direct",
        stamp_s: float = 0.0,
    ) -> ServoOutput:
        self.ensure_open()

        axis_norm = str(axis).strip().lower()
        output = self._servo_output(axis_norm, angle_deg)

        if axis_norm == "pan":
            self._state = PanTiltState(
                pan_deg=output.angle_deg,
                tilt_deg=self._state.tilt_deg,
                mode=mode,
                status=STATUS_DRYRUN if self.config.dryrun() else STATUS_OK,
                stamp_s=float(stamp_s),
                source=source,
            )
        elif axis_norm == "tilt":
            self._state = PanTiltState(
                pan_deg=self._state.pan_deg,
                tilt_deg=output.angle_deg,
                mode=mode,
                status=STATUS_DRYRUN if self.config.dryrun() else STATUS_OK,
                stamp_s=float(stamp_s),
                source=source,
            )
        else:
            raise ValueError(f"unknown head axis: {axis!r}")

        self._last_outputs = [output]
        return output

    def set_pan_tilt(
        self,
        pan_deg: int,
        tilt_deg: int,
        *,
        mode: str = MODE_MANUAL,
        source: str = "direct",
        stamp_s: float = 0.0,
    ) -> PanTiltState:
        self.ensure_open()

        limits = self.limits()
        pan = limits.clamp_pan(pan_deg)
        tilt = limits.clamp_tilt(tilt_deg)

        pan_output = self._servo_output("pan", pan)
        tilt_output = self._servo_output("tilt", tilt)

        self._last_outputs = [pan_output, tilt_output]
        self._state = PanTiltState(
            pan_deg=pan_output.angle_deg,
            tilt_deg=tilt_output.angle_deg,
            mode=mode,
            status=STATUS_DRYRUN if self.config.dryrun() else STATUS_OK,
            stamp_s=float(stamp_s),
            source=source,
        )

        return self._state

    def center(self, *, source: str = "center", stamp_s: float = 0.0) -> PanTiltState:
        return self.set_pan_tilt(
            self.calibration.pan.center_deg,
            self.calibration.tilt.center_deg,
            mode=MODE_CENTERING,
            source=source,
            stamp_s=stamp_s,
        )

    def hold(self, *, source: str = "hold", stamp_s: float = 0.0) -> PanTiltState:
        return PanTiltState(
            pan_deg=self._state.pan_deg,
            tilt_deg=self._state.tilt_deg,
            mode=self._state.mode,
            status=self._state.status,
            stamp_s=float(stamp_s),
            source=source,
        )

    def stop(self, *, source: str = "stop", stamp_s: float = 0.0) -> PanTiltState:
        self._state = PanTiltState(
            pan_deg=self._state.pan_deg,
            tilt_deg=self._state.tilt_deg,
            mode=MODE_IDLE,
            status=self._state.status,
            stamp_s=float(stamp_s),
            source=source,
        )
        return self._state

    def apply_command(self, command: PanTiltCommand) -> PanTiltState:
        cmd = command.normalized(self.limits())

        if cmd.command_type == COMMAND_STOP:
            return self.stop(source=cmd.source, stamp_s=cmd.stamp_s)

        if cmd == hold_command(source=cmd.source, stamp_s=cmd.stamp_s, reason=cmd.reason):
            return self.hold(source=cmd.source, stamp_s=cmd.stamp_s)

        target = cmd.target_from_state(self._state, self.limits())

        return self.set_pan_tilt(
            target.pan_deg,
            target.tilt_deg,
            mode=target.mode,
            source=cmd.source,
            stamp_s=cmd.stamp_s,
        )

    def apply_center_command(self, *, stamp_s: float = 0.0) -> PanTiltState:
        return self.apply_command(center_command(stamp_s=stamp_s))

    def status_dict(self) -> dict:
        return {
            "opened": self.opened,
            "backend": self.config.backend,
            "dryrun": self.config.dryrun(),
            "state": self.state.to_dict(),
            "last_outputs": [item.to_dict() for item in self.last_outputs],
        }


def make_dryrun_pantilt_driver(center_on_open: bool = False) -> PanTiltDriver:
    return PanTiltDriver(
        PanTiltDriverConfig(
            backend=BACKEND_DRYRUN,
            center_on_open=center_on_open,
            center_on_close=False,
        )
    )


def make_pca9685_pantilt_driver(center_on_open: bool = True, center_on_close: bool = True) -> PanTiltDriver:
    return PanTiltDriver(
        PanTiltDriverConfig(
            backend=BACKEND_PCA9685,
            center_on_open=center_on_open,
            center_on_close=center_on_close,
        )
    )


__all__ = [
    "BACKEND_PCA9685",
    "BACKEND_DRYRUN",
    "ServoOutput",
    "PanTiltDriverConfig",
    "PanTiltDriver",
    "make_dryrun_pantilt_driver",
    "make_pca9685_pantilt_driver",
]

File: savo_ws/src/core/savo_head/src/drivers/pantilt_driver.cpp
#include "savo_head/drivers/pantilt_driver.hpp"

#include <sstream>
#include <stdexcept>

namespace savo_head
{

PanTiltDriver::PanTiltDriver(PanTiltDriverConfig config)
: PanTiltDriver(
    config.normalized(),
    std::make_unique<Pca9685Driver>(config.normalized().pca9685_config()))
{
}

PanTiltDriver::PanTiltDriver(
  PanTiltDriverConfig config,
  std::unique_ptr<Pca9685Driver> pca_driver)
: config_(config.normalized()),
  calibration_(config_.calibration.normalized()),
  pca_driver_(std::move(pca_driver)),
  state_{
    calibration_.pan.center_deg,
    calibration_.tilt.center_deg,
    HeadMode::kIdle,
    driver_status_for_backend(config_.dryrun(), false),
    0.0,
    "init"}
{
  if (!pca_driver_) {
    pca_driver_ = std::make_unique<Pca9685Driver>(config_.pca9685_config());
  }
}

PanTiltDriver::~PanTiltDriver()
{
  try {
    close();
  } catch (...) {
  }
}

PanTiltDriver::PanTiltDriver(PanTiltDriver && other) noexcept
{
  *this = std::move(other);
}

PanTiltDriver & PanTiltDriver::operator=(PanTiltDriver && other) noexcept
{
  if (this == &other) {
    return *this;
  }

  try {
    close();
  } catch (...) {
  }

  config_ = other.config_;
  calibration_ = other.calibration_;
  pca_driver_ = std::move(other.pca_driver_);
  opened_ = other.opened_;
  state_ = other.state_;
  last_outputs_ = std::move(other.last_outputs_);
  last_error_ = std::move(other.last_error_);

  other.opened_ = false;
  other.last_error_.reset();

  return *this;
}

void PanTiltDriver::open()
{
  if (opened_) {
    return;
  }

  clear_error();

  try {
    pca_driver_->open();
    opened_ = true;

    update_state(
      state_.pan_deg,
      state_.tilt_deg,
      state_.mode,
      driver_status_for_backend(config_.dryrun(), opened_),
      CommandSource::kSystem,
      state_.stamp_s);

    if (config_.center_on_open) {
      const auto centered = center(CommandSource::kSystem, 0.0);
      (void)centered;
    }
  } catch (const std::exception & exc) {
    set_error(exc.what());
    opened_ = false;
    throw;
  }
}

void PanTiltDriver::close()
{
  if (opened_ && config_.center_on_close) {
    try {
      const auto centered = center(CommandSource::kSystem, state_.stamp_s);
      (void)centered;
    } catch (const std::exception & exc) {
      set_error(exc.what());
    }
  }

  if (pca_driver_) {
    pca_driver_->close();
  }

  opened_ = false;
  update_state(
    state_.pan_deg,
    state_.tilt_deg,
    HeadMode::kIdle,
    driver_status_for_backend(config_.dryrun(), opened_),
    CommandSource::kSystem,
    state_.stamp_s);
}

void PanTiltDriver::ensure_open()
{
  if (!opened_) {
    open();
  }
}

bool PanTiltDriver::opened() const
{
  return opened_;
}

bool PanTiltDriver::dryrun() const
{
  return config_.dryrun();
}

const PanTiltDriverConfig & PanTiltDriver::config() const
{
  return config_;
}

const HeadServoCalibration & PanTiltDriver::calibration() const
{
  return calibration_;
}

PanTiltLimits PanTiltDriver::limits() const
{
  return calibration_.limits();
}

PanTiltState PanTiltDriver::state() const
{
  return state_;
}

const std::vector<ServoOutput> & PanTiltDriver::last_outputs() const
{
  return last_outputs_;
}

std::optional<std::string> PanTiltDriver::last_error() const
{
  return last_error_;
}

Pca9685Driver & PanTiltDriver::pca9685()
{
  return *pca_driver_;
}

const Pca9685Driver & PanTiltDriver::pca9685() const
{
  return *pca_driver_;
}

void PanTiltDriver::clear_error()
{
  last_error_.reset();

  if (pca_driver_) {
    pca_driver_->clear_error();
  }
}

ServoOutput PanTiltDriver::set_axis(
  const std::string & axis,
  int angle_deg,
  HeadMode mode,
  CommandSource source,
  double stamp_s)
{
  ensure_open();

  try {
    const auto output = write_axis(axis, angle_deg);
    last_outputs_ = {output};

    if (axis == "pan") {
      update_state(
        output.angle_deg,
        state_.tilt_deg,
        mode,
        driver_status_for_backend(config_.dryrun(), opened_),
        source,
        stamp_s);
    } else if (axis == "tilt") {
      update_state(
        state_.pan_deg,
        output.angle_deg,
        mode,
        driver_status_for_backend(config_.dryrun(), opened_),
        source,
        stamp_s);
    } else {
      throw std::invalid_argument("unknown head axis: " + axis);
    }

    clear_error();
    return output;
  } catch (const std::exception & exc) {
    set_error(exc.what());
    throw;
  }
}

PanTiltState PanTiltDriver::set_pan_tilt(
  int pan_deg,
  int tilt_deg,
  HeadMode mode,
  CommandSource source,
  double stamp_s)
{
  ensure_open();

  try {
    const auto lim = limits();
    const auto pan = lim.clamp_pan(pan_deg);
    const auto tilt = lim.clamp_tilt(tilt_deg);

    const auto pan_output = write_axis("pan", pan);
    const auto tilt_output = write_axis("tilt", tilt);

    last_outputs_ = {pan_output, tilt_output};

    update_state(
      pan_output.angle_deg,
      tilt_output.angle_deg,
      mode,
      driver_status_for_backend(config_.dryrun(), opened_),
      source,
      stamp_s);

    clear_error();
    return state_;
  } catch (const std::exception & exc) {
    set_error(exc.what());
    throw;
  }
}

PanTiltState PanTiltDriver::center(CommandSource source, double stamp_s)
{
  return set_pan_tilt(
    calibration_.pan.center_deg,
    calibration_.tilt.center_deg,
    HeadMode::kCentering,
    source,
    stamp_s);
}

PanTiltState PanTiltDriver::hold(CommandSource source, double stamp_s)
{
  update_state(
    state_.pan_deg,
    state_.tilt_deg,
    state_.mode,
    driver_status_for_backend(config_.dryrun(), opened_),
    source,
    stamp_s);

  return state_;
}

PanTiltState PanTiltDriver::stop(CommandSource source, double stamp_s)
{
  update_state(
    state_.pan_deg,
    state_.tilt_deg,
    HeadMode::kIdle,
    driver_status_for_backend(config_.dryrun(), opened_),
    source,
    stamp_s);

  return state_;
}

PanTiltState PanTiltDriver::apply_command(const PanTiltCommand & command)
{
  if (command.type == CommandType::kStop) {
    return stop(command.source, command.stamp_s);
  }

  if (command.type == CommandType::kHold) {
    return hold(command.source, command.stamp_s);
  }

  const auto target = target_from_command(command, state_, limits());

  return set_pan_tilt(
    target.pan_deg,
    target.tilt_deg,
    target.mode,
    command.source,
    command.stamp_s);
}

std::string PanTiltDriver::status_text() const
{
  std::ostringstream stream;
  stream << "backend=" << config_.backend
         << ";opened=" << (opened_ ? "true" : "false")
         << ";dryrun=" << (dryrun() ? "true" : "false")
         << ";pan=" << state_.pan_deg
         << ";tilt=" << state_.tilt_deg
         << ";mode=" << to_string(state_.mode)
         << ";status=" << to_string(state_.status)
         << ";source=" << state_.source;

  if (last_error_.has_value() && !last_error_->empty()) {
    stream << ";error=" << *last_error_;
  }

  return stream.str();
}

ServoOutput PanTiltDriver::write_axis(const std::string & axis, int angle_deg)
{
  const auto output = angle_to_servo_output(axis, angle_deg, calibration_);
  const auto ticks = pca_driver_->set_servo_pulse_us(output.pca9685_channel, output.pulse_us);

  ServoOutput actual = output;
  actual.ticks = ticks;

  return actual;
}

void PanTiltDriver::set_error(std::string message)
{
  last_error_ = std::move(message);
  state_.status = HeadStatus::kError;
}

void PanTiltDriver::update_state(
  int pan_deg,
  int tilt_deg,
  HeadMode mode,
  HeadStatus status,
  CommandSource source,
  double stamp_s)
{
  state_ = PanTiltState{
    limits().clamp_pan(pan_deg),
    limits().clamp_tilt(tilt_deg),
    mode,
    status,
    stamp_s,
    to_string(source)
  }.normalized(limits());
}

}  // namespace savo_head
